File: src-tauri/src/commands/backup.rs
use tauri::Manager;
use tauri_plugin_dialog::DialogExt;
use crate::auth::{crypto, state as auth_state};
use crate::db;
use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Key, Nonce,
};
use rand::{rngs::OsRng, RngCore};
use zeroize::Zeroize;
use std::io::{Write as _};
use std::path::PathBuf;

// .ptdata file magic + version
const MAGIC: &[u8; 4] = b"PTDT";
const VERSION: u32 = 2;  // v2: plain SQLite (no SQLCipher layer)

#[tauri::command]
pub async fn pick_backup_folder(app: tauri::AppHandle) -> Result<Option<String>, String> {
    let folder = app
        .dialog()
        .file()
        .set_title("Select Backup Folder")
        .blocking_pick_folder();
    match folder {
        Some(path) => Ok(Some(path.to_string())),
        None => Ok(None),
    }
}

/// Export the database to a .ptdata file encrypted with the given password.
#[tauri::command]
pub fn export_data(
    app: tauri::AppHandle,
    pin: String,
    password: String,
    dest_path: String,
) -> Result<(), String> {
    // Verify PIN (security gate — user must prove identity before exporting)
    let app_dir = app.path().app_data_dir().map_err(|e| e.to_string())?;
    let auth_path = app_dir.join("auth.json");
    let auth_json = std::fs::read_to_string(&auth_path).map_err(|e| e.to_string())?;
    let auth_config: crypto::AuthConfig = serde_json::from_str(&auth_json).map_err(|e| e.to_string())?;
    crypto::unwrap_key(&auth_config.pin_wrapped, &pin)
        .map_err(|_| "Incorrect PIN".to_string())?;

    // Read the plain SQLite file directly
    let db_path = app_dir.join("portfolio.db");
    let plaintext_bytes = std::fs::read(&db_path).map_err(|e| e.to_string())?;

    // Encrypt with a fresh random export key (AES-256-GCM)
    let mut export_key = [0u8; 32];
    OsRng.fill_bytes(&mut export_key);

    let mut db_nonce_bytes = [0u8; 12];
    OsRng.fill_bytes(&mut db_nonce_bytes);

    let db_cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&export_key));
    let db_nonce = Nonce::from_slice(&db_nonce_bytes);
    let encrypted_db = db_cipher
        .encrypt(db_nonce, plaintext_bytes.as_ref())
        .map_err(|e| format!("DB encryption failed: {e}"))?;

    // Wrap export key with export password (Argon2id)
    let wrapped_export_key = crypto::wrap_key(&export_key, &password)
        .map_err(|e| e.to_string())?;
    export_key.zeroize();

    // Write .ptdata file
    let salt_bytes = base64_decode(&wrapped_export_key.salt)?;
    let wrap_nonce_bytes = base64_decode(&wrapped_export_key.nonce)?;
    let wrap_ct_bytes = base64_decode(&wrapped_export_key.ciphertext)?;

    let mut file = std::fs::File::create(&dest_path).map_err(|e| e.to_string())?;
    file.write_all(MAGIC).map_err(|e| e.to_string())?;
    file.write_all(&VERSION.to_le_bytes()).map_err(|e| e.to_string())?;
    file.write_all(&salt_bytes).map_err(|e| e.to_string())?;
    file.write_all(&wrap_nonce_bytes).map_err(|e| e.to_string())?;
    let wct_len = wrap_ct_bytes.len() as u32;
    file.write_all(&wct_len.to_le_bytes()).map_err(|e| e.to_string())?;
    file.write_all(&wrap_ct_bytes).map_err(|e| e.to_string())?;
    file.write_all(&db_nonce_bytes).map_err(|e| e.to_string())?;
    let db_len = encrypted_db.len() as u64;
    file.write_all(&db_len.to_le_bytes()).map_err(|e| e.to_string())?;
    file.write_all(&encrypted_db).map_err(|e| e.to_string())?;
    file.flush().map_err(|e| e.to_string())?;

    Ok(())
}

/// Import a .ptdata file — wipes current DB and replaces with the imported one.
#[tauri::command]
pub fn import_data(
    app: tauri::AppHandle,
    password: String,
    src_path: String,
) -> Result<(), String> {
    let app_dir = app.path().app_data_dir().map_err(|e| e.to_string())?;

    // Read .ptdata file
    let file_bytes = std::fs::read(&src_path).map_err(|e| e.to_string())?;
    let mut cursor = 0usize;

    if file_bytes.get(cursor..cursor + 4) != Some(MAGIC) {
        return Err("Invalid file format — not a .ptdata file".to_string());
    }
    cursor += 4;

    let version = u32::from_le_bytes(
        file_bytes[cursor..cursor + 4].try_into().map_err(|_| "Corrupt file")?
    );
    cursor += 4;
    if version != VERSION {
        return Err(format!("Unsupported .ptdata version {version} (expected {VERSION})"));
    }

    let salt_bytes: [u8; 16] = file_bytes[cursor..cursor + 16]
        .try_into().map_err(|_| "Corrupt file: salt")?;
    cursor += 16;

    let wrap_nonce_bytes: [u8; 12] = file_bytes[cursor..cursor + 12]
        .try_into().map_err(|_| "Corrupt file: wrap nonce")?;
    cursor += 12;

    let wct_len = u32::from_le_bytes(
        file_bytes[cursor..cursor + 4].try_into().map_err(|_| "Corrupt file: wct len")?
    ) as usize;
    cursor += 4;
    let wrap_ct_bytes = file_bytes[cursor..cursor + wct_len].to_vec();
    cursor += wct_len;

    let db_nonce_bytes: [u8; 12] = file_bytes[cursor..cursor + 12]
        .try_into().map_err(|_| "Corrupt file: db nonce")?;
    cursor += 12;

    let db_len = u64::from_le_bytes(
        file_bytes[cursor..cursor + 8].try_into().map_err(|_| "Corrupt file: db len")?
    ) as usize;
    cursor += 8;
    let encrypted_db = &file_bytes[cursor..cursor + db_len];

    // Unwrap export key using password
    use base64::{engine::general_purpose::STANDARD as B64, Engine};
    let wrapped = crypto::WrappedKey {
        salt: B64.encode(salt_bytes),
        nonce: B64.encode(wrap_nonce_bytes),
        ciphertext: B64.encode(&wrap_ct_bytes),
    };
    let mut export_key = crypto::unwrap_key(&wrapped, &password)
        .map_err(|_| "Incorrect export password".to_string())?;

    // Decrypt the db blob
    let db_cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&export_key));
    let db_nonce = Nonce::from_slice(&db_nonce_bytes);
    let plaintext_bytes = db_cipher
        .decrypt(db_nonce, encrypted_db)
        .map_err(|_| "Decryption failed — file may be corrupt or wrong password".to_string())?;
    export_key.zeroize();

    // Write plaintext to temp file then atomically replace the DB
    let db_path = app_dir.join("portfolio.db");
    let temp_path = app_dir.join("import_temp.db");

    std::fs::write(&temp_path, &plaintext_bytes).map_err(|e| {
        format!("Failed to write temp file: {e}")
    })?;

    // Remove stale WAL/SHM files
    let _ = std::fs::remove_file(app_dir.join("portfolio.db-wal"));
    let _ = std::fs::remove_file(app_dir.join("portfolio.db-shm"));

    std::fs::rename(&temp_path, &db_path).map_err(|e| {
        let _ = std::fs::remove_file(&temp_path);
        format!("Failed to replace database: {e}")
    })?;

    db::replace(&db_path)?;

    Ok(())
}

fn base64_decode(s: &str) -> Result<Vec<u8>, String> {
    use base64::{engine::general_purpose::STANDARD as B64, Engine};
    B64.decode(s).map_err(|e| format!("base64 decode error: {e}"))
}

File: backend/routers/backup.py
"""Backup export/import — port of src-tauri/src/commands/backup.rs.

.ptdata binary format (must stay byte-compatible with Rust):
  "PTDT"              4 bytes  magic
  version: u32 LE    4 bytes  = 2
  salt                16 bytes (Argon2 salt for wrapping the export key)
  wrap_nonce          12 bytes
  wrap_ct_len: u32 LE 4 bytes
  wrap_ciphertext     variable
  db_nonce            12 bytes
  db_len: u64 LE      8 bytes
  encrypted_db        variable
"""
import os
import sqlite3
import struct
from pathlib import Path
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel

from core import auth as auth_module
from core import session as session_module
from routers.deps import get_conn

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

router = APIRouter(tags=["backup"])
Conn = Annotated[sqlite3.Connection, Depends(get_conn)]

_MAGIC   = b"PTDT"
_VERSION = 2


class ExportInput(BaseModel):
    pin: str
    password: str
    dest_path: str


class ImportInput(BaseModel):
    password: str
    src_path: str


def _require_session(request: Request) -> bytes:
    token = request.headers.get("X-Session-Token")
    if not token:
        raise HTTPException(status_code=401, detail="Missing X-Session-Token header")
    try:
        return session_module.get_key(token)
    except KeyError:
        raise HTTPException(status_code=401, detail="Invalid session token")


@router.post("/export")
def export_data(body: ExportInput, request: Request, conn: Conn):
    app_dir: Path = request.app.state.app_dir

    # Verify PIN
    try:
        auth_module.login(app_dir, body.pin)
    except ValueError:
        raise HTTPException(status_code=401, detail="Incorrect PIN")

    # Read the plain DB file (connection is still open — checkpoint first)
    conn.execute("PRAGMA wal_checkpoint(TRUNCATE)")
    conn.commit()

    db_path = app_dir / "portfolio.db"
    plaintext_db = db_path.read_bytes()

    # Encrypt DB with a fresh random export key
    export_key = os.urandom(32)
    db_nonce   = os.urandom(12)
    encrypted_db = AESGCM(export_key).encrypt(db_nonce, plaintext_db, None)

    # Wrap export key with the user-supplied export password (Argon2id + AES-GCM)
    wrapped = auth_module.wrap_key(export_key, body.password)

    import base64
    salt_bytes     = base64.b64decode(wrapped["salt"])
    wrap_nonce     = base64.b64decode(wrapped["nonce"])
    wrap_ct        = base64.b64decode(wrapped["ciphertext"])

    # Write binary file
    with open(body.dest_path, "wb") as f:
        f.write(_MAGIC)
        f.write(struct.pack("<I", _VERSION))        # u32 LE
        f.write(salt_bytes)                          # 16 bytes
        f.write(wrap_nonce)                          # 12 bytes
        f.write(struct.pack("<I", len(wrap_ct)))     # u32 LE
        f.write(wrap_ct)
        f.write(db_nonce)                            # 12 bytes
        f.write(struct.pack("<Q", len(encrypted_db)))# u64 LE
        f.write(encrypted_db)

    return {"ok": True}


@router.post("/import")
def import_data(body: ImportInput, request: Request, conn: Conn):
    import base64

    app_dir: Path = request.app.state.app_dir

    data = Path(body.src_path).read_bytes()
    cur  = 0

    if data[cur:cur + 4] != _MAGIC:
        raise HTTPException(status_code=400, detail="Invalid file — not a .ptdata file")
    cur += 4

    version = struct.unpack_from("<I", data, cur)[0]; cur += 4
    if version != _VERSION:
        raise HTTPException(status_code=400, detail=f"Unsupported .ptdata version {version}")

    salt       = data[cur:cur + 16]; cur += 16
    wrap_nonce = data[cur:cur + 12]; cur += 12
    wct_len    = struct.unpack_from("<I", data, cur)[0]; cur += 4
    wrap_ct    = data[cur:cur + wct_len]; cur += wct_len
    db_nonce   = data[cur:cur + 12]; cur += 12
    db_len     = struct.unpack_from("<Q", data, cur)[0]; cur += 8
    encrypted_db = data[cur:cur + db_len]

    # Unwrap export key
    wrapped = {
        "salt":       base64.b64encode(salt).decode(),
        "nonce":      base64.b64encode(wrap_nonce).decode(),
        "ciphertext": base64.b64encode(wrap_ct).decode(),
    }
    try:
        export_key = auth_module.unwrap_key(wrapped, body.password)
    except ValueError:
        raise HTTPException(status_code=401, detail="Incorrect export password")

    # Decrypt DB
    try:
        plaintext_db = AESGCM(export_key).decrypt(db_nonce, encrypted_db, None)
    except Exception:
        raise HTTPException(status_code=400, detail="Decryption failed — corrupt file or wrong password")

    # Close current connection, write new DB, reopen
    master_key = _require_session(request)

    conn.close()
    request.app.state.conn = None

    db_path   = app_dir / "portfolio.db"
    temp_path = app_dir / "import_temp.db"

    temp_path.write_bytes(plaintext_db)
    for suffix in ("-wal", "-shm"):
        try:
            (app_dir / f"portfolio.db{suffix}").unlink()
        except FileNotFoundError:
            pass
    temp_path.rename(db_path)

    # Re-encrypt the newly placed DB under the current master key
    from core import db as db_module
    new_conn = db_module.open_database(app_dir, master_key)
    request.app.state.conn = new_conn

    return {"ok": True}
